pkg/volume/csi/csi_driver_reg.go

```diff
@@ -0,0 +1,280 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// Name of node annotation that contains JSON map of driver names to node
+	// names
+	annotationKey = "csi.volume.kubernetes.io/nodeid"
+)
+
+type csiDriver struct {
+	driverName     string
+	driverEndpoint string
+}
+
+// csiDrivers map keep track of all registered CSI drivers on the node and their
+// corresponding sockets
+var csiDrivers sync.Map
+
+// labelManagement is struct of channels used for communication between the driver registration
+// code and the go routine responsible for managing the node's labels
+type labelManager struct {
+	in  chan csiDriver
+	out chan bool
+}
+
+var lm labelManager
+
+// nodeLabelManager waits for labeling requests initiated by the driver's registration
+// process.
+// TODO (sbezverk) Right now the driver registration only registers drivers, not clear
+// how to de-register the driver when CSI driver is terminated.
+// TODO (sbezverk) Check for pluginwatcher behavior on kubelet restart, CSI would require
+// replaying registrations for all existing socket. (Possible issue is when kubelet is up and tryig
+// to replay registrations but corresponding driver-registrars are not up yet, would it keep retrying?)
+func nodeLabelManager(nodeName types.NodeName, k8s kubernetes.Interface) {
+	for driver := range lm.in {
+		glog.Infof("><SB> Request to update node %s's label for CSI driver %s", nodeName, driver.driverName)
+		err := verifyAndAddNodeId(string(nodeName), k8s.CoreV1().Nodes(), driver.driverName, string(nodeName))
+		if err != nil {
+			glog.Errorf("Failed to update node %s's annotation with error: %+v", nodeName, err)
+			lm.out <- false
+			continue
+		}
+		lm.out <- true
+	}
+}
+
+// RegistrationCallback is called by kubelet's plugin watcher upon detection
+// of a new registration socket opened by CSI Driver registrar side car.
+func RegistrationCallback(pluginName string, endpoint string, versions []string, socketPath string) (error, chan bool) {
+
+	glog.Infof("><SB> Callback from kubelet with plugin name: %s endpoint: %s versions: %s socket path: %s",
+		pluginName, endpoint, strings.Join(versions, ","), socketPath)
+
+	// Storing endpoint of newly registered CSI driver into the map, where CSI driver name will be the key
+	// all other CSI components will be able to get the actual socket of CSI drivers by its name.
+	csiDrivers.Store(pluginName, csiDriver{driverName: pluginName, driverEndpoint: endpoint})
```

Comment:
Also `csiDrivers` is not used anywhere. Next step is to modify all the existing code to use `csiDrivers` to find the socket instead of generating it manually if `enablePluginsWatcher` is true.